File: packages/rac-server/rules_visualizer_rac/references.py
"""Attach references.json policy-doc citations onto Model nodes.

Separate from the parser because it's a generic post-processing step that
works on any Model regardless of source format (old `.rac` or new RuleSpec).
"""

from __future__ import annotations

import json
from pathlib import Path
from typing import Any


def resolve_references(model: dict[str, Any], ruleset_dir: str) -> None:
    """Load `<ruleset_dir>/references.json` and attach resolved
    {section, document} entries onto each mapped node's `references` field.
    Silently no-ops if the file is missing or malformed."""
    ref_path = Path(ruleset_dir) / "references.json"
    if not ref_path.is_file():
        return

    try:
        refs = json.loads(ref_path.read_text(encoding="utf-8"))
    except Exception as e:
        print(f"  Warning: failed to parse references.json: {e}")
        return

    docs_by_id = {d["id"]: d for d in refs.get("documents", [])}
    sections_by_id = {s["id"]: s for s in refs.get("sections", [])}

    mappings_by_path: dict[str, list[str]] = {}
    for m in refs.get("mappings", []):
        mappings_by_path.setdefault(m["nodePath"], []).append(m["sectionId"])

    for node in model.get("nodes", {}).values():
        node_name = node.get("name", "")
        section_ids = mappings_by_path.get(node_name)
        if not section_ids:
            continue
        resolved = []
        for section_id in section_ids:
            section = sections_by_id.get(section_id)
            if not section:
                continue
            doc = docs_by_id.get(section.get("documentId", ""))
            if not doc:
                continue
            resolved.append({"section": section, "document": doc})
        if resolved:
            node["references"] = resolved

File: packages/factgraph-core/src/references.ts
import fs from 'node:fs'
import path from 'node:path'
import type {
  Model,
  PolicyReferences,
  ResolvedReference,
} from 'rules-visualizer-shared-types'

/**
 * Load `references.json` from a ruleset directory and attach resolved policy
 * citations onto the matching model nodes. Silently no-ops when the file is
 * absent or malformed — references are an optional enhancement, not a
 * requirement.
 *
 * `references.json` schema lives in `rules-visualizer-shared-types`. Nodes are
 * matched by path (Fact Graph nodes use their `path` as the node name).
 */
export function resolveReferences(model: Model, rulesetDir: string): void {
  const refPath = path.join(rulesetDir, 'references.json')
  if (!fs.existsSync(refPath)) return

  let refs: PolicyReferences
  try {
    refs = JSON.parse(fs.readFileSync(refPath, 'utf-8'))
  } catch (e) {
    console.warn(`  Warning: failed to parse references.json: ${e}`)
    return
  }

  const docsById = new Map(refs.documents.map((d) => [d.id, d]))
  const sectionsById = new Map(refs.sections.map((s) => [s.id, s]))

  const mappingsByPath = new Map<string, string[]>()
  for (const m of refs.mappings) {
    const list = mappingsByPath.get(m.nodePath) ?? []
    list.push(m.sectionId)
    mappingsByPath.set(m.nodePath, list)
  }

  for (const node of Object.values(model.nodes)) {
    const nodePath = node.name // FG nodes use path as name
    const sectionIds = mappingsByPath.get(nodePath)
    if (!sectionIds) continue

    const resolved: ResolvedReference[] = []
    for (const sectionId of sectionIds) {
      const section = sectionsById.get(sectionId)
      if (!section) continue
      const document = docsById.get(section.documentId)
      if (!document) continue
      resolved.push({ section, document })
    }
    if (resolved.length > 0) {
      node.references = resolved
    }
  }
}
